Debian packaging lintian clean

paulvt opened the pull request:
This PR patches up the Debian/Ubuntu packaging as to be valid with respect to the latest Debian packaging standards by means of manual inspection/review (as a Debian developer) and the lintian tool.

paulvt committed:
Fix empty paragraph in description; remove article from short description

paulvt committed:
Fix permissions of the DLL files and game executables

The DLL files should not be exectable, nor should the game executables
be writeable by the group.

paulvt committed:
Create a link for the OpenRA server manpage; compress the normal one

The command-line arguments of the openra-server executable are the
same as for openra.

paulvt committed:
Fix up the changelog

Just use a dummy changelog entry about the release and fix
a syntax error in the entry footer.

The changelog should not contain the contents of Changelog.md as this
results in an unparsable Debian changelog.  The game's changelog is already
available as CHANGELOG.html in /usr/share/doc/openra.

paulvt committed:
Add a lintian override for the font duplication

This will probably never be fixed and the override is there to document
that this is ignored explicitly.

paulvt committed:
Fix the issues with filenames for the md5sums

Lintian reported missing md5sums for all files under usr/ and reported the
existence of md5sums for all files under /usr that we're not in the
package.

pchote commented:
> The command-line arguments of the openra-server executable are the
same as for openra.

While this is *technically* true, OpenRA.Server.exe will parse and then ignore everything except Game.Mod and Server.*.

pchote commented:
Given that (a) the majority of the arguments for `openra` aren't listed in the man page (because they don't have `[Desc]`s), and (b) it's kind of useless, since they're all exposed ingame anyway, I think it would be much better if `--man-page` was changed to generate a full / complete page for `openra-server` and we write a static stub page for `openra`.

paulvt commented:
I could change the missing man page for the server into an override so that it left as a to-do item.
It could also be added as a note that the server only accepts certain parameters.
But I agree, both executables probably deserve a better man page. It is however nothing I can change given that I wanted this to be purely a PR that touches on packaging.

pchote commented:
Adding an override and opening a ticket (ideally blocking the next release) to fix the man page properly sounds good to me.

paulvt committed:
Leave the openra-server man page out for now; add an lintian override

See also issue #12525.

Mailaender commented on packaging/linux/deb/openra.lintian-overrides:
```diff
@@ -0,0 +1,3 @@
+openra: duplicate-font-file usr/lib/openra/mods/common/FreeSans.ttf also in fonts-freefont-ttf
+openra: duplicate-font-file usr/lib/openra/mods/common/FreeSansBold.ttf also in fonts-freefont-ttf
+openra: binary-without-manpage usr/games/openra-server
```
These are valid Lintian complaints. I suggest to not disable them.

paulvt replied:
They are, and they are not disabled but ignored for good/temporary reasons; lintian still lists them for me.

If you can confirm that it can find the fonts if they're not at that location and if you are willing to add the depend on `fonts-freefont-ttf`, then we can rid of the first two.
The third warning is discussed above, it will be unignored once the man-page is created (#12525).

I choose to go for the overrides because they document that things are left unclean for a specific reason. It might be a good idea to document the reason though.

pchote replied:
I guess you could symlink the font files, but that feels like a nastier hack than keeping everything self contained with a bit of duplication.

Mailaender replied:
Well, this is a thirdparty not a distribution package so duplicating is better, but why do you turn off (or what else does override mean) this warnings?

paulvt replied:
Symlinking of fonts and JS files etc. actually happens a lot, but that is usually in a situation that upstream doesn't want to change things. As you say @Mailaender, we don't have a distribution package here, so might as well leave it like that.
I'd like to have it suppress the warning while documenting it, so if things become unclean again in the future (lintian-wise), it is noticed immediately and either fixed or documented and overridden (in very rare cases).

I am proceeding with adding some comments and then I'm quite happy with the result!

Mailaender replied:
Okay, fair enough. I think we don't see any Lintian output as our CI scripts don't run it. Thanks for making us aware this way.

paulvt committed:
Add comments to the overrides; drop the missing manpage override

The missing openra-server manpage warning should not be overriden,
as the manpage should be there. (See also: #12525).

pchote commented:
@obrakmann: Can you please give this a test?  I've (hopefully only temporarily...) misplaced the drive with my Ubuntu VM.

obrakmann commented:
Package builds fine, but I get a lot of warnings for file permissions when I run lintian on the package:
```
$ lintian -c /tmp/openra_test.20170121_all.deb
W: openra: non-standard-file-perm usr/lib/openra/Eluant.dll.config 0664 != 0644
W: openra: non-standard-file-perm usr/lib/openra/OpenAL-CS.dll.config 0664 != 0644
W: openra: non-standard-file-perm usr/lib/openra/SDL2-CS.dll.config 0664 != 0644
W: openra: non-standard-file-perm ... use --no-tag-display-limit to see all (or pipe to a file/program)
W: openra: non-standard-executable-perm usr/lib/openra/launch-dedicated.sh 0775 != 0755
W: openra: non-standard-executable-perm usr/lib/openra/launch-game.sh 0775 != 0755
W: openra: non-standard-dir-perm usr/lib/openra/lua/ 0775 != 0755
W: openra: non-standard-dir-perm usr/lib/openra/mods/cnc/ 0775 != 0755
W: openra: non-standard-dir-perm usr/lib/openra/mods/cnc/audio/ 0775 != 0755
W: openra: non-standard-dir-perm ... use --no-tag-display-limit to see all (or pipe to a file/program)
W: openra: binary-without-manpage usr/games/openra-server
N: 2 tags overridden (2 warnings)
```

If I use `--no-tag-display-limit`, it complains about basically all the files. Is there maybe a difference between Debian and Ubuntu?

paulvt commented:
Can you produce a file/dir listing of your package? The complaints about the permissions mentioned in your output have exactly the correct permissions it wants them to be in my .deb file?

obrakmann commented:
[Yup, here it is](https://github.com/OpenRA/OpenRA/files/721440/deb.files.txt)

paulvt commented:
I wonder, since this package is constructed without tools and such, if it is due to a difference in umask or something? Mine is 0022. Because all manually created and chmod'ed stuff by the `buildpackage.sh` script is fine in your listing, but the remaining, just copied files, all have the group write bit set for example.

obrakmann commented:
Good call, my umask is 0002

paulvt commented:
Maybe we should just add `umask 0022` even to the `package-all.sh` script, because when distributing you might not set the user up with group writable files because you don't know what their setup is.
I can add that to the pull request if you guys desire so (and it fixes it for you @obrakmann).

obrakmann commented:
No, setting the umask before packaging doesn't help, as `cp` preserves the mode from files in the repo, which were created with the old umask.

pchote commented:
would `chmod -R`ing the target directory after copying it be too big of a foot gun?

paulvt commented:
Absolutely right, yeah.
Maybe it's better to switch to `install` instead of using `cp`. Might make the whole thing cleaner and we use the intended tool.

pchote commented:
@paulvt: [package-all.sh#L33](https://github.com/OpenRA/OpenRA/blob/bleed/packaging/package-all.sh#L33) have fun :smile:

obrakmann commented:
I'm fine with either way. `chmod`ing probably takes less work, and produces a clean package (just tested).

paulvt commented:
Ok, will have a look at it tomorrow. Thanks for the review/check.

paulvt committed:
Nothing should have group writable permissions

These might occur due to difference in umask values.

paulvt commented:
I went with the "chmod -R" solution, because I'm afraid I destroy things when using `install`, although that is still the cleaner solution.  Please check if the remaining warnings are resolved now.

obrakmann reviewed (approved):
Yup, it's fine now, 👍

pchote reviewed (approved):
Looks good to me, so with @obrakmann's testing this gets my :+1: too.  If anything breaks we can deal with it at playtest time.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/66c9fc073c7726a388a93e26a44ae80393715ce6
